test/e2e/util.go

```diff
@@ -454,6 +448,91 @@ func newRequestFromForm(form *html.Node, currentURL *url.URL, user string) (*htt
 	return req, nil
 }
 
+// Varying the login name for each test ensures we test a fresh grant
+func createTestIdP(
+	t *testing.T,
+	kubeClient *kubernetes.Clientset,
+	oauthClient configv1client.OAuthInterface,
+	userClientSet *userclients.Clientset,
+	nsName string,
+	numUsers int,
+) ([]string, func()) {
+	oauthConfig, err := oauthClient.Get(context.TODO(), "cluster", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	var users []string
+	for i := 0; i < numUsers; i++ {
+		users = append(users, fmt.Sprintf("testuser%d", i))
+	}
+
+	kubeClient.CoreV1().Secrets("openshift-config").Create(context.TODO(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nsName + "htpasswd",
+			},
+			Data: map[string][]byte{
+				"htpasswd": []byte(strings.Join(users, ":$2y$05$0Fk2s.0FbLy0FZ82JAqajOV/kbT/wqKX5/QFKgps6J69J2jY6r5ZG\n")), // bcrypt of 'password'
+			},
+		},
+		metav1.CreateOptions{},
+	)
+
+	oauthConfig.Spec.IdentityProviders = append(
+		oauthConfig.Spec.IdentityProviders, configv1.IdentityProvider{
+			Name: nsName,
+			IdentityProviderConfig: configv1.IdentityProviderConfig{
+				Type: "HTPasswd",
+				HTPasswd: &configv1.HTPasswdIdentityProvider{
+					FileData: configv1.SecretNameReference{
+						Name: nsName + "htpasswd",
+					},
+				},
+			},
+		},
+	)
+
+	_, err = oauthClient.Update(context.TODO(), oauthConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+
+	cleanup := func() {
+		err := kubeClient.CoreV1().Secrets("openshift-config").Delete(context.TODO(), nsName+"htpasswd", metav1.DeleteOptions{})
+		if err != nil {
+			t.Logf("failed to delete secret openshift-config/%shtpasswd: %v", nsName, err)
+		}
+		oauthConfig, err := oauthClient.Get(context.TODO(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			t.Logf("failed to get the oauth/cluster config during cleanup: %v", err)
+		}
+		for i := range oauthConfig.Spec.IdentityProviders {
+			if providers := oauthConfig.Spec.IdentityProviders; providers[i].Name == nsName {
+				oauthConfig.Spec.IdentityProviders = deleteProvider(providers, i)
+				break
+			}
+		}
+		_, err = oauthClient.Update(context.TODO(), oauthConfig, metav1.UpdateOptions{})
+		if err != nil {
+			t.Logf("failed to remove the test IdP from oauth/cluster: %s", err)
+		}
+
+		for i := 0; i < numUsers; i++ {
+			username := fmt.Sprintf("testuser%d", i)
```

Comment:
`username` is `users[i]`?

Reply:
I'd rather keep it like this in case `users`, reference to which we return, were tampered with:
https://play.golang.org/p/oHB69CMloDv